dali/base/dadfs.cpp

```diff
@@ -7929,6 +7930,130 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+{
+    StringBuffer superName;
+    Owned<IDistributedSuperFile> superfile;
+
+    if (!tree->hasProp("@name"))
+        throw MakeStringException(-1, "createSuperFile: There is no SuperFile name in XML");
+
+    tree->getProp("@name", superName);
+    if (superName.isEmpty())
+        throw MakeStringException(-1, "createSuperFile: SuperFile name is empty in XML");
+
+    const char *pSuperName = superName.str();
+    if (queryDistributedFileDirectory().exists(pSuperName, user, false, true))
+        throw MakeStringException(-1, "createSuperFile: SuperFile '%s' already exists!", pSuperName);
+
+    // Check Physical Parts of SubFiles
+    unsigned numSubfiles = tree->getPropInt("@numsubfiles", 0);
+    if (numSubfiles)
+    {
+        Owned<IPropertyTreeIterator> treeIter = tree->getElements("SubFile");
+        ForEach(*treeIter)
+        {
+            IPropertyTree & st = treeIter->get();
+            StringBuffer subfilename;
+            st.getProp("@name", subfilename);
+            if (!queryDistributedFileDirectory().existsPhysical(subfilename.str(), user))
+                throw MakeStringException(-1, "createSuperFile: Physical Parts of '%s' not found!", subfilename.str());
+        }
+    }
+    // It seems XML is fine add this superfile to root tree
+
+    // Get the correct branch name
+    StringBuffer branchName;
+    branchName.set(queryDfsXmlBranchName(DXB_SuperFile)).trim();
+    const char *branch = branchName.str();
+
+    // Get the head path and connect with it
+    CDfsLogicalFileName sdlname;
+    sdlname.set(pSuperName);
+    StringBuffer head;
+    sdlname.makeScopeQuery(head.clear(), true);
+    Owned<IRemoteConnection> conn = querySDS().connect(head.str(), myProcessSession(), 0, FOREIGN_DALI_TIMEOUT);
+    if (!conn)
+        throw MakeStringException(-1, "createSuperFile: Could not connect to '%s'", head.str());
+
+    Owned<IPropertyTree> root = conn->getRoot();
+    root->addPropTree(branch, LINK(tree));
+    conn->commit();
+    PROGLOG("SuperFile '%s' loaded from XML", pSuperName);
+    conn->close();
+
+    // Update sub-files
+    StringAttr supNam;
+    sdlname.normalizeName(pSuperName, supNam, true );
+    superName.set(supNam.str());
+
+    unsigned subnum = 0;
+    if (numSubfiles)
+    {
+        Owned<IPropertyTreeIterator> treeIter = tree->getElements("SubFile");
+        ForEach(*treeIter)
+        {
+            IPropertyTree & st = treeIter->get();
+            StringBuffer subfilename;
+            st.getProp("@name", subfilename);
+            CDfsLogicalFileName sublname;
+            sublname.set(subfilename.str());
+            if (!sublname.isExternal() && !sublname.isForeign())
+            {
+                StringBuffer subquery;
+                sublname.makeFullnameQuery(subquery, DXB_File, true);
+                Owned<IRemoteConnection> subconn = querySDS().connect(subquery.str(), myProcessSession(), RTM_LOCK_WRITE, FOREIGN_DALI_TIMEOUT);
```

Comment:
should use existing DFS methods as much as possible.
So you could 'lookup' here, rather than connect directly (which would bypass locking etc. etc.)